src/plugins/elasticsearch_ui_shared/public/request.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { useEffect, useState, useRef } from 'react';
+
+export interface SendRequest {
+  path: string;
+  method: string;
+  body?: any;
+  uimActionType?: string;
+}
+
+export interface SendRequestResponse {
+  data: any;
+  error: Error;
+}
+
+export interface UseRequest extends SendRequest {
+  interval?: number;
+  initialData?: any;
+  processData?: any;
+}
+
+export const sendRequest = async (
+  httpClient: any,
+  { path, method, body }: SendRequest
+): Promise<Partial<SendRequestResponse>> => {
+  try {
+    // NOTE: This is tightly coupled to Angular's $http service.
+    const response = await httpClient[method](path, body);
+
+    if (typeof response.data === 'undefined') {
+      throw new Error(response.statusText);
+    }
+
+    return { data: response.data };
+  } catch (e) {
+    return {
+      error: e.response ? e.response : e,
+    };
+  }
+};
+
+export const useRequest = (
+  httpClient: any,
+  { path, method, body, interval, initialData, processData }: UseRequest
+) => {
+  // Main states for tracking request status and data
+  const [error, setError] = useState<null | any>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [data, setData] = useState<any>(initialData);
+
+  // Consumers can use isInitialRequest to implement a polling UX.
+  const [isInitialRequest, setIsInitialRequest] = useState<boolean>(true);
```

Comment:
What do you think to have a counter instead of a boolean? So it can be used in different scenarios (solve the current problem, but also know how many requests have been made).

```js
const [totalRequests, setTotalRequests] = useState<number>(0);
```

Reply:
I'm not opposed to this, but do you have a use case in mind? Otherwise it seems like YAGNI.